components/omega/OmegaBuild.cmake

```diff
@@ -77,61 +82,261 @@ macro(preset)
+      if (_CXX_VER_RESULT EQUAL 0)
+
+        string(REGEX MATCH "HIP|hip"       _HIP_CHECK "${_CXX_VER_OUTPUT}")
+        string(REGEX MATCH "AMD|amd"       _AMD_CHECK "${_CXX_VER_OUTPUT}")
+        string(REGEX MATCH "NVCC|nvcc"     _NVCC_CHECK "${_CXX_VER_OUTPUT}")
+        string(REGEX MATCH "NVIDIA|nvidia" _NVIDIA_CHECK "${_CXX_VER_OUTPUT}")
+
+        if(_HIP_CHECK AND _AMD_CHECK)
+          set(OMEGA_ARCH "HIP")
+
+        elseif(_NVCC_CHECK AND _NVIDIA_CHECK)
+          set(OMEGA_ARCH "CUDA")
+
+        else()
+          set(OMEGA_ARCH "OPENMP")
+
+        endif()
+      else()
+        set(OMEGA_ARCH "OPENMP")
+
+      endif()
     endif()
   endif()
 
-  set(OMEGA_C_COMPILER ${_OMEGA_C_COMPILER})
-  set(CMAKE_C_COMPILER ${OMEGA_C_COMPILER})
+  message(STATUS "OMEGA_ARCH = ${OMEGA_ARCH}")
 
-  set(OMEGA_CXX_COMPILER ${_OMEGA_CXX_COMPILER})
-  set(CMAKE_CXX_COMPILER ${OMEGA_CXX_COMPILER})
+  # create a profile script
+  set(_ProfileScript ${OMEGA_BUILD_DIR}/omega_profile.sh)
+  file(WRITE ${_ProfileScript}  "#!/usr/bin/env bash\n\n")
+  file(APPEND ${_ProfileScript} "source ./omega_env.sh\n\n")
+  file(APPEND ${_ProfileScript} "# modify 'OUTFILE' with a path in that the profiler can\n")
+  file(APPEND ${_ProfileScript} "# create files such as a path in a scratch file system.\n")
 
-  set(OMEGA_Fortran_COMPILER ${_OMEGA_Fortran_COMPILER})
+  # set C and Fortran compilers *before* calling CMake project()
+  set(CMAKE_C_COMPILER ${OMEGA_C_COMPILER})
   set(CMAKE_Fortran_COMPILER ${OMEGA_Fortran_COMPILER})
 
-  message(STATUS "OMEGA_C_COMPILER = ${OMEGA_C_COMPILER}")
-  message(STATUS "OMEGA_CXX_COMPILER = ${OMEGA_CXX_COMPILER}")
-  message(STATUS "OMEGA_Fortran_COMPILER = ${OMEGA_Fortran_COMPILER}")
-  message(STATUS "MPI_EXEC = ${MPI_EXEC}")
+# TODO: do we want to use these variables?
+#  # Set compiler and linker flags
+#  if (CXXFLAGS)
+#    separate_arguments(_CXXFLAGS NATIVE_COMMAND ${CXXFLAGS})
+#    list(APPEND OMEGA_CXX_FLAGS ${_CXXFLAGS})
+#  endif()
+#
+#  if (LDFLAGS)
+#    separate_arguments(_LDFLAGS NATIVE_COMMAND ${LDFLAGS})
+#    list(APPEND OMEGA_LINK_OPTIONS ${_LDFLAGS})
+#  endif()
+#
+#  if (SLIBS)
+#    separate_arguments(_SLIBS NATIVE_COMMAND ${SLIBS})
+#    list(APPEND OMEGA_LINK_OPTIONS ${_SLIBS})
+#  endif()
+
+  # set CXX compiler *before* calling CMake project()
+  if(OMEGA_ARCH STREQUAL "CUDA")
+
+    if(NOT OMEGA_CUDA_COMPILER)
+      find_program(OMEGA_CUDA_COMPILER
+        "nvcc_wrapper"
+        PATHS "${OMEGA_SOURCE_DIR}/../../externals/ekat/extern/kokkos/bin"
+      )
+    endif()
+
+    if(OMEGA_CUDA_COMPILER)
+      message(STATUS "OMEGA_CUDA_COMPILER = ${OMEGA_CUDA_COMPILER}")
+
+    else()
+      message(FATAL_ERROR "Cuda compiler is not found." )
+    endif()
+
+    set(CMAKE_CXX_COMPILER ${OMEGA_CUDA_COMPILER})
+    set(CMAKE_CUDA_HOST_COMPILER ${OMEGA_CXX_COMPILER})
+
+    # overwrite CMAKE_CXX_FLAGS and CMAKE_EXE_LINKER_FLAGS defined in
+    # cime configuration because those could break CUDA build
```

Comment:
I'm curious, what breaks in cuda builds?

Reply:
@bartgol , when I used "CMAKE_EXE_LINKER_FLAGS" as defined in E3SM CIME, I got an error during cmake configuration due to the compiler flags(-acc -gpu=cc70,cc60 -Minfo=accel) in "CMAKE_EXE_LINKER_FLAGS".

/global/homes/y/youngsun/repos/github/Omega/externals/ekat/extern/kokkos/bin/nvcc_wrapper  -ccbin /opt/cray/pe/craype/2.7.20/bin/CC -Wno-deprecated-gpu-targets -DKokkos_ENABLE_SERIAL=On   -Wl,--allow-multiple-definition -lstdc++ -acc -gpu=cc70,cc60 -Minfo=accel  CMakeFiles/cmTC_1feb7.dir/testCXXCompiler.cxx.o -o cmTC_1feb7
nvc++-Error-More than one output file will overwrite /tmp/tmpxft_0006ef5a_00000000-6_cmTC_1feb7_dlink.o

In the above log, CC wrapps nvidia hpc compiler.

This issue is related to your another comment related to using "mpicxx" instead of "nvcc_wrapper", and I will write about it in the comment section.